lib/evmone/eof.cpp

```diff
@@ -90,37 +151,261 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+            return EOFValidationError::inputs_outputs_num_above_limit;
+
+        if (t.max_stack_height > MAX_STACK_HEIGHT)
+            return EOFValidationError::max_stack_height_above_limit;
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
-    size_t i = 0;
-    uint8_t op = code[0];
-    while (i < code.size())
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
+    for (size_t i = 0; i < code.size(); ++i)
     {
-        op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        const auto op = code[i];
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
-        ++i;
-    }
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 < code.size())
+            {
+                const auto count = code[i + 1];
+                if (count < 1)
+                    return EOFValidationError::invalid_rjumpv_count;
+                i += static_cast<size_t>(1 /* count */ + count * 2 /* tbl */);
+            }
+            else
+                return EOFValidationError::truncated_instruction;
+        }
+        else
+            i += instr::traits[op].immediate_size;
 
-    if (!instr::traits[op].is_terminating)
-        return EOFValidationError::missing_terminating_instruction;
+        if (i >= code.size())
+            return EOFValidationError::truncated_instruction;
+    }
 
     return EOFValidationError::success;
 }
 
+bool validate_rjump_destinations(bytes_view code) noexcept
+{
+    // Collect relative jump destinations and immediate locations
+    const auto code_size = code.size();
+    std::vector<size_t> rjumpdests;
+    std::vector<bool> immediate_map(code_size);
+
+    for (size_t i = 0; i < code_size; ++i)
+    {
+        const auto op = code[i];
+
+        if (op == OP_RJUMP || op == OP_RJUMPI)
+        {
+            const auto offset = read_int16_be(&code[i + 1]);
+            const auto jumpdest = static_cast<int32_t>(i) + 3 + offset;
+            if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                return false;
+            rjumpdests.push_back(static_cast<size_t>(jumpdest));
+        }
+        else if (op == OP_RJUMPV)
+        {
+            constexpr auto REL_OFFSET_SIZE = sizeof(int16_t);
+
+            const auto count = code[i + 1];
+
+            const auto post_offset = 1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */;
+
+            for (size_t k = 0; k < count * REL_OFFSET_SIZE; k += REL_OFFSET_SIZE)
+            {
+                const auto rel_offset = read_int16_be(&code[i + 1 + 1 + static_cast<uint16_t>(k)]);
+                const auto jumpdest = static_cast<int>(i + post_offset) + rel_offset;
```

Comment:
This is using `int`, while `RJUMP(I)` is using `int32_t` for the internal calculation. Probably `int32_t` is more consistent from the perspective of EIP-1985.